internal/gh/rest.go

```
@@ -0,0 +1,164 @@
+
+    return req, nil
+}
+
+func GetForcePushInfo(owner string, repo string, token string, branch string) (int, error){
+/*
+    Entry point for the Force Push processing, setting the url, query paramenters and creating the client.
+*/
+    slog.Default().Info("Getting repo info - getForcePushInfo method")
+    repoActivityUrl := baseURL + "/repos/" + owner + "/" + repo + "/activity"
+    queryParameters := map[string]string{
+        "per_page" : "100",
+        "activity_type" : "force_push",
+        "ref" : branch,
+    }
+
+    client := &http.Client{}
+
+    err := processForcePushRequest(client, repoActivityUrl, token, queryParameters) 
+    if err != nil {
+        slog.Default().Info("Getting repo info falied - %v getForcePushInfo method", err)
+        return 0, err
+    }    
+    return NumberForcePush, nil
+}
+
+func executeHTTPRequest(client *http.Client, req *http.Request) (*http.Response, error){
+/*
+    Execute the HTTP request and return response
+*/
+
+    resp, err := client.Do(req)
+	RequestCounter++
+	if err != nil {
+        return nil, err
+	}
+    return resp, nil
+}
+
+func processForcePushRequest(client *http.Client, repoActivityUrl string, token string, queryParameters map[string] string) error {
+/*
+    Method where calls to other methods (createHTTPRequest, executeHTTPRequest, processHttpResponse) is handled and loop is added so that requests are processed till there no rel = "next"
+*/
+
+    slog.Default().Info("Getting Force Push Info - processForcePushRequestAndResponse method")
+    
+    for true{
+        httpReq, err := createHttpRequest(client, repoActivityUrl, token, queryParameters)
+        if err != nil {
+	        slog.Default().Error("Failed to create HTML request: %v", err)
+	        return  err
+        }
+        resp, err := executeHTTPRequest(client, httpReq)
+        if err != nil {
+		    slog.Default().Error("Failed to execute HTML request: %v - processForcePushRequest method", err)
+		    return err
```

this breaks the loop if one request fails. however, maybe a follow up request will work again. Failure of one request should not terminate the whole loop

---

How do you want me to handle this, ignore the error and only display it via slog or store the err message and then concatenate each time we have an error  and return it once (This could change how I handle the error in other methods and in the processor class)?

---

start with logging it. The idiomatic solution is usually to return an array of errors for multi step functions where every step can fail individually.
However, in this case we won't handle the individual errors anyways so it's ok to just log them. Maybe add a counter so we can print how many times the operation failed in the end so we get an easy overview

---

There is one more problem with this. Unless we execute the request, we don't know if there is a next page available or what is the URL of the next page. Both those are only available after the request is executed and the response is processed. Since the response headers contain those information.

---

Ok, so if the method only returns an error iff we can not get a next URL then it's fine to terminate the whole loop